File: src/lib/ferret/matrix.clj
(defn row-count [^matrix u]
  "return obj<number>(matrix::row_count(u));")

(defn column-count [^matrix u]
  "return obj<number>(matrix::column_count(u));")

(defn zeros [^size_t r, ^size_t c]
  "return obj<matrix_t>(matrix::zeros(r, c));")

(defn ones [^size_t r, ^size_t c]
  "return obj<matrix_t>(matrix::ones(r,c));")

(defn full [^size_t r, ^size_t c, ^real_t v]
  "return obj<matrix_t>(matrix::ones(r,c,v));")

(defn eye [^size_t n]
  "return obj<matrix_t>(matrix::eye(n));")

(defn add [^matrix u ^matrix v]
  "return obj<matrix_t>(u + v);")

(defn sub [^matrix u ^matrix v]
  "return obj<matrix_t>(u - v);")

(defn mul [^matrix u ^real_t s]
  "return obj<matrix_t>(u * s);")

(defn mmul [^matrix u ^matrix v]
  "return obj<matrix_t>(u * v);")

(defn equals [^matrix u ^matrix v]
  "if (u == v)
       return cached::true_o;
   else
       return cached::false_o;")

(defn normalise [^matrix u]
  "return obj<matrix_t>(matrix::normalise(u));")

(defn norm [^matrix u]
  "return obj<number>(matrix::norm_euclidean(u));")

(defn mset! [^matrix u ^size_t r ^size_t c ^real_t v]
  "u(r, c, v);")

(defn mget [^matrix u ^size_t r ^size_t c]
  "return obj<number>(u(r, c));")

(defn cout [^matrix u]
  "std::cout << u;")

(defn seq [u]
  "return obj<array_seq<real_t, number>>(u);")

File: test/runtime/matrix.cpp
#include <cassert>
#include <runtime.h>

int main() {
  using namespace ferret;
    
  //initializers
  assert((matrix::zeros(2,2)  == matrix::into<2,2>(0,0,0,0)));
  assert((matrix::ones(2,2)   == matrix::into<2,2>(1,1,1,1)));
  assert((matrix::eye(2)      == matrix::into<2,2>(1,0,0,1)));
  assert((matrix::full(2,2,4) == matrix::into<2,2>(4,4,4,4)));

  //shape
  assert((matrix::row_count(matrix::zeros(2,3))    == 2));
  assert((matrix::column_count(matrix::zeros(2,3)) == 3));

  //operations
  matrix ones  = matrix::ones(2,2);
  matrix zeros = matrix::zeros(2,2);
  matrix twos  = matrix::full(2,2,2);

  assert((ones - ones == zeros));
  assert((ones + ones == twos));
  assert((ones * 2    == twos));

  auto v3d = matrix::into<1,3>(0, 10, 0);

  assert((matrix::norm_euclidean(v3d) == 10));
  assert((matrix::normalise(v3d)      == matrix::into<1,3>(0, 1, 0)));
  return 0;
}
